### DISCUSSION TITLE
Any plans for officially supporting what symfony/flex and php-http/discovery does?

### DISCUSSION THREAD

**Discussion by mxr576** [composer/composer#11392] [Q&A / Support]
Both `symfony/flex` and `php-http/discovery` interacts with Composer in a special way. They both manipulates the list of required dependencies and for that they hacks their way through to get access to the original installer object... and they also try to preserve configurations supplied to the original Composer command that was executed and who knows what else... :exploding_head: 

e.g,: https://github.com/php-http/discovery/blob/1.15.2/src/Composer/Plugin.php

I haven't finished with my analyses but `wikimedia/composer-merge-plugin` could do something similar way before these other plugins were born. 

I believe there is a real need for manipulating the list of required packages from a plugin - like instead of suggesting packages to be installed, actually installing them when XY package is installed. I wonder if there are plans to make writing these kind of plugins easier or not.

**Update 2023.03.22 17:40 CET**: I'd probably also add `laminas/laminas-dependency-plugin` to the list because it also performs dependency manipulation.

**Comment by mxr576** [comment]
@nicolas-grekas Hi, if you do not mind, as the author the `php-http/discovery`, I'd summon you here to share your experience about 2 cents about writing that Composer plugin. Do/did you have any maintenance concerns for that plugin?

**Reply by nicolas-grekas** [reply]
I agree it'd be nice to have some clean API to do what is done here:
https://github.com/php-http/discovery/blob/1.15.2/src/Composer/Plugin.php#L146-L184

**Reply by Seldaek** [reply]
Can you try to explain what you are trying to achieve? I find it hard to parse the layers of hacks here..

Is it trying to run a second update? After a first update was done? Or after the install step is completed?

- symfony/flex restricts the amount of visible packages but this can now be done cleanly via the `pre-pool-create` event (https://getcomposer.org/doc/articles/scripts.md#plugin-events) 
- the laminas plugin I don't know why it's only compatible with Composer 2.2 and not newer ones, but that doesn't really motivate me to look further into it
- php-http/discovery is trying to add some requirements and that is a really special use case, I don't know how we could serve this better, but would need more details as per my question above..

**Reply by nicolas-grekas** [reply]
This is not related to version constraints. What flex does here is taking packages listed in a special "flex-require" entry in composer.json and putting them in "require". (it does so to be sure that its constraints are applied even on first installs).
For php-http/discovery, it adds an entry to composer.json when nothing implements one of the supported `*-implementation` virtual packages and it restarts "composer up" basically.

**Reply by mxr576** [reply]
> pre-pool-create: occurs before the Pool of packages is created, and lets you filter the list of packages that is going to enter the Solver.

**filter**... can it extend the list with new dependencies that should be installed?

Practically speaking I am playing with two ideas in my head to serve certain internal needs related to "better automate project setups", in other words, reduce cognitive load for installing and configuring packages the right way :tm:. Depending on which idea seems more feasible I can start working out a solution...

One of them is similar to what Flex does (if I understand it right), when package X is requested, instead of installing package X the dependencies of package X gets installed and tracked in the root composer.json and package X can also execute some tasks to "terraform" the project where it got installed.

The other one is automatically installing additional dependencies when package Y is installed and if necessary also performing additional "terraforming" tasks. Of course, when additional packages should be installed I would like to be sure that the dependency resolver performs a regular dependency resolving and installs the accurate version.

So I think in short I am probably looking for solutions for manipulating requirements before they get passed to the dependency resolver (adding packages, replacing packages... maybe removing) and performing additional tasks defined by a newly installed package when it is installed - without marking it as a `composer-plugin` (so I need another `composer-plugin` that reacts on the dependency install and executes tasks).

**Reply by stof** [reply]
Well, when you say "similar to what Flex does", you need to be more precise. Flex does at least 10 _different_ things

**Reply by nicolas-grekas** [reply]
In this context we have here, I'd say packs auto-unpacking, which I forgot to mention before.

**Reply by Seldaek** [reply]
Here is a whole set of different things IMO.. 

- dependency restarting is fairly simple, we could perhaps add a special exception class or smth that if thrown would be caught and would restart the update.. So you could update the composer.json and throw and done. I guess that'd help @nicolas-grekas ?
- manipulating requirements on the fly if not in composer.json is a bit nastier IMO, you could do it via the pre-pool-create by modifying the package metadata in memory but I can't say I love the idea..
- executing post-pkg-install hooks sounds also possible with a plugin for sure, I don't think that you need anything new for that

Am I missing anything?
